internal/config/migration/multi_account.go

```diff
@@ -0,0 +1,210 @@
+// allow forward compatability for older versions of gh and also to avoid
+// breaking existing users of go-gh which looks at a specific location
+// in the config for oauth tokens that are stored insecurely.
+
+type MultiAccount struct {
+	// Allow injecting a transport layer in tests.
+	Transport http.RoundTripper
+}
+
+func (m MultiAccount) PreVersion() string {
+	return ""
+}
+
+func (m MultiAccount) PostVersion() string {
+	return "1"
+}
+
+func (m MultiAccount) Do(c *config.Config) error {
+	hostnames, err := c.Keys(hostsKey)
+	// [github.com, github.localhost]
+	// We wouldn't expect to have a hosts key when this is the first time anyone
+	// is logging in with the CLI.
+	var keyNotFoundError *config.KeyNotFoundError
+	if errors.As(err, &keyNotFoundError) {
+		return nil
+	}
+	if err != nil {
+		return CowardlyRefusalError{errors.New("couldn't get hosts configuration")}
+	}
+
+	// If there are no hosts then it doesn't matter whether we migrate or not,
+	// so lets avoid any confusion and say there's no migration required.
+	if len(hostnames) == 0 {
+		return nil
+	}
+
+	// Otherwise let's get to the business of migrating!
+	for _, hostname := range hostnames {
+		token, inKeyring, err := getToken(c, hostname)
+		if err != nil {
+			return CowardlyRefusalError{fmt.Errorf("couldn't find oauth token for %q: %w", hostname, err)}
+		}
+
+		username, err := getUsername(c, hostname, token, m.Transport)
+		if err != nil {
+			return CowardlyRefusalError{fmt.Errorf("couldn't get user name for %q: %w", hostname, err)}
+		}
+
+		if err := migrateConfig(c, hostname, username); err != nil {
+			return CowardlyRefusalError{fmt.Errorf("couldn't not migrate config for %q: %w", hostname, err)}
+		}
+
+		if err := migrateToken(hostname, username, token, inKeyring); err != nil {
+			return CowardlyRefusalError{fmt.Errorf("couldn't not migrate oauth token for %q: %w", hostname, err)}
+		}
+	}
+
+	return nil
+}
+
+func getToken(c *config.Config, hostname string) (string, bool, error) {
+	if token, _ := c.Get(append(hostsKey, hostname, "oauth_token")); token != "" {
+		return token, false, nil
+	}
+	token, err := keyring.Get(keyringServiceName(hostname), "")
+	if err != nil {
+		return "", false, err
+	}
+	if token == "" {
+		return "", false, errors.New("token not found in config or keyring")
+	}
+	return token, true, nil
+}
+
+func getUsername(c *config.Config, hostname, token string, transport http.RoundTripper) (string, error) {
+	username, _ := c.Get(append(hostsKey, hostname, "user"))
+	if username != "" && username != "x-access-token" {
+		return username, nil
+	}
+	opts := ghAPI.ClientOptions{
+		Host:      hostname,
+		AuthToken: token,
+		Transport: transport,
+	}
+	client, err := ghAPI.NewGraphQLClient(opts)
+	if err != nil {
+		return "", err
+	}
+	var query struct {
+		Viewer struct {
+			Login string
+		}
+	}
+	err = client.Query("CurrentUser", &query, nil)
+	return query.Viewer.Login, err
```

Comment:
Maybe a bit of a nitpick but I find this pattern to be hard to understand because I'm never sure whether there is supposed to be meaning in the first return value. If it's ok I'd much prefer an `if err != nil { return "", nil }` and then `return query.Viewer.Login, nil`.